pkg/ipcache/ipcache.go

```diff
@@ -422,8 +422,17 @@ func (ipc *IPCache) upsertLocked(
 		namedPortsChanged = namedPortsChanged && ipc.needNamedPorts.Load()
 	}
 
-	if hostIP != nil {
-		hostID = ipc.AllocateNodeID(hostIP)
+	if oldHostIP.Equal(hostIP) {
+		if hostIP != nil {
+			hostID, _ = ipc.GetNodeID(hostIP)
+		}
+	} else {
+		if hostIP != nil {
+			hostID = ipc.AllocateNodeID(hostIP)
+		}
+		if oldHostIP != nil {
+			ipc.DeallocateNodeID(oldHostIP)
+		}
```

Comment:
I missing the steps to get two nodeIDs. Considering this flow,

First we get a pod update]
    PodUpdate -> upsertLocked(hostIP) -> AllocateNodeID(hostIP)  -> creates hostIP:idA:1

this allocates new ID 'idA' with refcnt 1 and maps it to hostIP. Then in example we get nodeUpdate.

    NodeUpdate->deleteNodeIPs->enableIPsecIPv4->allocateIDForNode

So I think the concern is the deleteNodeIPs in case 2 when a podupdate happens first. (just trying to be sure I understand) For deleteNodeIPs to be called we need an 'oldNode != nil' If this is the case then we didn't come from NodeAdd() we must have come from NodeUpdate.

Can we assume Node events for a single node are ordered so that NodeAdd always happens before NodeUpdate. I think so. Then there shouldn't be a problem because the NodeAdd will increment the refcnt and the following NodeUpdate could decrement it but that would be from 2->1.

So the case 2 would be: podUpdate -> NodeAdd -> NodeUpdate?

Please correct me I think I missed something in your comment?

Reply:
> So the case 2 would be: podUpdate -> NodeAdd -> NodeUpdate?

Case 2 is NodeAdd -> PodAdd -> PodUpdate -> NodeUpdate. We only care about the later two.

---

Let me keep the syntax from the other thread:
- `OldIP` is the IP address that was removed from node A and later reused by node B.
- `NewIP` is the IP address that was added to node A, to replace OldIP.
- `OldNodeID` is the node ID allocated for OldIP.

All is as seen from a node C, different from nodes A and B.

0. Initial state on node C.
    - <ins>State:</ins> `OldNodeID` has refcount 3 (1 for the node, +2 for two pods).
2. A pod update is received with `OldIP` -> `NewIP`.
    - We will decrement refcount for `OldNodeID`. 
    - We will allocate a new node ID for `NewIP`. Let's call it `NewNodeID`.
    - <ins>State:</ins> `OldNodeID` has refcount 2. `NewNodeID` has refcount 1.
    - After this pod update, I believe we will already have XfrmOutPolBlock drops because the ipcache for this pod is updated but the XFRM config isn't. Is that wrong?
3. A node update for node A is received with `OldIP` -> `NewIP`.
    - We will unmap `OldIP` from `OldNodeID`. 
    - We will call `allocateIDForNode(NodeA)`. That will call `getNodeIDForNode(NodeA)`. Nodes have several IP addresses. At this point, node A will have NewIP associated to `NewNodeID`, but it will also have other IP addresses still associated to `OldNodeID`. Which node ID `getNodeIDForNode(NodeA)` will pick is random because it relies on a map iteration.
    - Let's consider `OldNodeID` is picked. Then, in `allocateIDForNode`, all IP addresses of the node, including `NewIP` will be associated to `OldNodeID`. We will also increment the refcount for `OldNodeID`.
    - The XFRM config will remain the same since `OldNodeID` was selected.
    - <ins>State:</ins> `OldNodeID` has refcount 3 (!!). `NewNodeID` has refcount 1.
    - At this point, I believe the ipcache entry for the pod that was updated is still using `NewNodeID` while the XFRM config is using `OldNodeID`. That will result in XfrmOutPolBlock drops.
    - In addition, the refcount for `OldNodeID` is now incorrect (3 vs 1 node + 1 pod) and `OldNodeID` will thus never be deallocated.

We have big issues already, but let's keep going :sweat_smile: 
Let's introduce node B.

4. Node B is created with OldIP as one of its IP addresses.
    - We will call `allocateIDForNode(NodeB)` as part of the node creation. That will map all IPs of node B, including `OldIP`, to |OldNodeID` since the mapping OldIP <> OldNodeID still exists.
    - We will install XFRM configs for node B. Those configs will use `OldNodeID`. Small difference between ENI and overlay here: on ENI, we will overwrite the XFRM configs for node A. On overlay we will install additional entries that use the same node ID but different CIDRs.
    - <ins>State:</ins> `OldNodeID` has refcount 4. `NewNodeID` has refcount 1.
    - Both node A and B are using `OldNodeID` in XFRM configs, but the ipcache entries are a mix of both node IDs.
5. On the creation of a pod on node B.
    - We will create ipcache entries for all pods of node B. They will be using `OldNodeID`.
    - <ins>State:</ins> `OldNodeID` has refcount 5. `NewNodeID` has refcount 1.